Add schema validation

TomasJani opened the pull request:
Schema validation on types string, bool and user-friendly representation.

jpopelka commented on packit/config/package_config.py:
```diff
@@ -254,6 +254,84 @@ def __eq__(self, other: object):
+    def format_validation_errors(self):
+        if not self.schema_primitive_errors:
+            return (
+                f"{self.package_config.config_file_path} is valid and ready to bu used"
```
```suggestion
                f"{self.package_config.config_file_path} is valid and ready to be used"
```

jpopelka commented on packit/config/package_config.py:
```diff
@@ -254,6 +254,84 @@ def __eq__(self, other: object):
+                "create_pr": package_config.create_pr,
+                "actions": package_config.actions,
+            }
+        )
```
Hasn't the `package_config` at this point been already validated ?
([get_packit_api()](https://github.com/packit-service/packit/blob/master/packit/cli/utils.py#L111) -> [get_local_package_config()](https://github.com/packit-service/packit/blob/8833635fdfbc681a518f75b7a1a363785a319ee2/packit/config/package_config.py#L296) -> [parse_loaded_config()](https://github.com/packit-service/packit/blob/8833635fdfbc681a518f75b7a1a363785a319ee2/packit/config/package_config.py#L360) -> [get_from_dict()](https://github.com/packit-service/packit/blob/8833635fdfbc681a518f75b7a1a363785a319ee2/packit/config/package_config.py#L176))

Anyway, something like `PackageConfigSchema().dump(package_config)` should create the dict for you.

jpopelka commented on tests/integration/test_config_validation.py:
```diff
@@ -0,0 +1,87 @@
+)
+def test_schema_validation_primitive_types(packit_config, contains):
+    validator = PackageConfigValidationRepresentation(packit_config)
+    assert str(validator).find(contains) != -1
```
```suggestion
    assert contains in str(validator)
```

TomasJani commented:
Thank you for review.

> Hasn't the package_config at this point been already validated ?
(get_packit_api() -> get_local_package_config() -> parse_loaded_config() -> get_from_dict())

Yes, for what I understood, point of this is to create readable output for user.

> Anyway, something like PackageConfigSchema().dump(package_config) should create the dict for you.

Does not work because function _serialize in ActionField (schema.py) is not implemented. And i don't want to validate every field in PackageConfig. Fields that have List or Nested types need to be treated differently.

TomasTomecek commented:
I am trying this on packit itself and am getting:
```
Neither `specfile_path` nor `downstream_package_name` set.
```
Which is fine, we have defaults for those.

I am trying now on a different repo and am getting the same message even though specfile_path is set.

TomasTomecek commented on packit/config/package_config.py:
```diff
@@ -254,6 +254,151 @@ def __eq__(self, other: object):
         )
```
You are adding too much new classes with no docstrings - for me it's easy to get lost there: since the 4 new classes accept PackageConfig in a constructor, I'd personally turn them into a single one. On the other hand, feel free to explain what's your thinking process for the current design and we can discuss.

lachmanfrantisek commented on packit/cli/validate_config.py:
```diff
@@ -0,0 +1,52 @@
+@cover_packit_exception
+def validate_config(config, path_or_url):
+    """
+    Validate PackageConfig - check if specfile_path and synced_files are present and
+    if downstream_package_name is present in Fedora
```
What about some general sentence or a list of checks to be able to easily update that docstring?

It's important since this docstring is used in the CLI in the `--help` message.

lachmanfrantisek commented on packit/config/package_config.py:
```diff
@@ -254,6 +254,151 @@ def __eq__(self, other: object):
+        )
+        self.schema_list_errors: str = str(ListValidationError(package_config))
+
+    def get_validation_errors_str(self) -> str:
+        if not (self.schema_primitive_errors or self.schema_list_errors):
+            return f"{self.config_file_path} is valid and ready to be used"
+
+        validation_errors = f"{self.config_file_path} does not pass validation:\n"
+        validation_errors += self.schema_primitive_errors
+        validation_errors += self.schema_list_errors
+
+        return validation_errors
+
+    def __repr__(self) -> str:
+        return self.get_validation_errors_str()
```
I would maybe differentiate between `__repr__`/`__str__` and output used for users.

- `__str__`/`__repr__` something like `PackageConfigValidation(config_file_path='the/path/to/config.yaml)` or `PackageConfigValidation(config_file_path='the/path/to/config.yaml, errors=['error1', 'error2'])`
  - Developer-focused view, I would not use those for forwarding user output. Somer property for accessing errors should be enough and the formatting can be done elsewhere.
- output printed to the user:
```
We found the following errors:
- error 1
- error 2
```
  - What about implementing that in the CLI method and not here? (All used attributes can be used from outside of the class as well.) It allows us easily update/tweak/colour the output.


Same can be applied to other classes.

lachmanfrantisek commented on packit/config/package_config.py:
```diff
@@ -254,6 +254,151 @@ def __eq__(self, other: object):
+class PackageConfigValidation:
+    def __init__(self, package_config: PackageConfig):
+        self.config_file_path: Optional[str] = package_config.config_file_path
+        self.schema_primitive_errors: str = str(
+            PrimitiveValidationError(package_config)
+        )
+        self.schema_list_errors: str = str(ListValidationError(package_config))
```
Since those two attributes need to be calculated, what aboutcreating [a property](https://www.freecodecamp.org/news/python-property-decorator/) for them instead.

lachmanfrantisek commented on packit/config/package_config.py:
```diff
@@ -254,6 +254,151 @@ def __eq__(self, other: object):
+        return error.format(self.field_name, type(self.field_value).__name__)
+
+
+class PrimitiveValidationError:
```
I am a bit confused about the naming. Classes that end with `Error` or `Exceptions` are usually errors or exceptions. (And can be used in the `try-except`.)

csomh commented on packit/cli/validate_config.py:
```diff
@@ -0,0 +1,52 @@
+    if downstream_package_name is present in Fedora
+    """
+    api = get_packit_api(config=config, local_project=path_or_url)
+    validation = PackageConfigValidation(api.package_config)
```
I find doing all the work in the constructor very confusing and out of pattern for how Python code is usually written. Other commands seem to call the API at this point. It would make sense to see something similar here too.

csomh commented on packit/config/package_config.py:
```diff
@@ -254,6 +254,151 @@ def __eq__(self, other: object):
+
+        return PackageConfigSchema().validate(
+            {
+                "config_file_path": package_config.config_file_path,
```
Why do primitives and lists be treated differently?

csomh commented on packit/config/package_config.py:
```diff
@@ -254,6 +254,151 @@ def __eq__(self, other: object):
+
+
+class ValidationError:
+    default_error_messages = {
```
How do the errors raised by Marshmallow look like? Could we use those instead of crafting new ones? I'm a little bit uncomfortable maintaining a list of error messages here, as it most probably will become a pothole of bugs.

TomasJani commented:
Thank all of you for review,
I tried to adjust my code according to all the comments.
Hope it's better then before.

csomh commented:
> Build failed.
> 
> * [packit-pre-commit ](https://softwarefactory-project.io/zuul/t/local/build/846fccfe7e9b42a9ac9ed2bec61326f5) : FAILURE in 2m 29s
> * [packit-tests-rpm ](https://softwarefactory-project.io/zuul/t/local/build/9852604ca6204c16bd966c5f2dd936f9) : FAILURE in 8m 35s

@TomasJani what about these failures?

TomasJani commented:
> @TomasJani what about these failures? 

I need to purposely put incorrect values according to mypy into PackageConfig to test if my validation detects them correctly. 

Test failures are caused by `PackageConfigSchema().validate()` on Marshmallow 2, it does not return a dictionary of validation errors like Marshmallow 3, but throws ValidationError.

TomasTomecek commented:
for the mypy errors, this has worked for me: suffix the problematic line with `# type: ignore`

> Test failures are caused by `PackageConfigSchema().validate()` on Marshmallow 2, it does not return a dictionary of validation errors like Marshmallow 3, but throws ValidationError.

Hm, this is an interesting problem: I can see 2 ways out of it:
1. Also support the interface of marsh 2 - catch that exception, process it and produce a nice error message
2. Disable this functionality with marsh 2

Since we support packit on Fedora versions which are still on marsh 2, I'd like to for the first option - catch the exception, make sense out of it, and probably turn it into a dict and then just continue with the current code path.

What do others think?

TomasTomecek commented:
2 more comments:
1. this PR needs rebase
2. If everything is valid, I'd probably print to stdout that the config is valid (since packit is pretty verbose by default)

TomasJani commented:
Thank you for review.
Now, it should be passing.

csomh commented on packit/config/package_config.py:
```diff
@@ -254,6 +255,85 @@ def __eq__(self, other: object):
+        try:
+            config = self.package_config
+            return PackageConfigSchema().validate(
+                {
+                    "config_file_path": config.config_file_path,
+                    "specfile_path": config.specfile_path,
+                    "downstream_package_name": config.downstream_package_name,
+                    "upstream_project_url": config.upstream_project_url,
+                    "upstream_package_name": config.upstream_package_name,
+                    "upstream_ref": config.upstream_ref,
+                    "upstream_tag_template": config.upstream_tag_template,
+                    "dist_git_base_url": config.dist_git_base_url,
+                    "dist_git_namespace": config.dist_git_namespace,
+                    "spec_source_id": config.spec_source_id,
+                    "create_pr": config.create_pr,
+                    "actions": config.actions,
+                    "create_tarball_command": config.create_tarball_command,
+                    "current_version_command": config.current_version_command,
+                    "allowed_gpg_keys": config.allowed_gpg_keys,
+                    "patch_generation_ignore_paths": config.patch_generation_ignore_paths,
+                }
```
Is this the data to be validated? Shouldn't this be `self.package_config` instead?

TomasTomecek commented on packit/cli/validate_config.py:
```diff
@@ -0,0 +1,58 @@
+    - checks missing values
+    - checks incorrect types
+
+    PATH_OR_URL argument is a local path or a URL to the upstream git repository,
```
I'd say this is just a copy-paste, but I agree with Hunor, let's rephrase to
```
PATH_OR_URL argument is a local path or a URL to a git repository with packit configuration file
```

TomasTomecek commented:
Nice, tests are passing, well done!

The last test is failing on all PRs and we are looking into it.

Please address Hunor's comments and I think that we're good to merge, awesome work! :beer:

jpopelka commented:
recheck

TomasJani commented:
> Is this the data to be validated? Shouldn't this be self.package_config instead?

I'm not sure if I understand. I renamed self.package_config to config because lines underneath were too long. I refactored validated dictionary into variable to solve this problem in latest commit. 
Everything else should be fixed.

csomh commented:
> > Is this the data to be validated? Shouldn't this be self.package_config instead?
> 
> I'm not sure if I understand. I renamed self.package_config to config because lines underneath were too long. I refactored validated dictionary into variable to solve this problem in latest commit.
> Everything else should be fixed.

Thanks! Might be that I missed that renaming :man_shrugging: 

I have concern regarding that mapping in `validate_primitive()` (which probably also applies to the other `validate_*` methods: if there is a change in `PackageConfig` this mapping will need to be also updated.

One way to overcome this could be to implement `get()` on `PackageConfig`, like:

```python
     def get(self, value, default=None):
         return self.__dict__.get(value, default)
```

This way, the mapping wouldn't be needed (and we wouldn't be required to update it in the future when `PackageConfig` changes) and you could just call `validate(self.package_config)`.

@TomasTomecek how do you feel about the above?

csomh commented on packit/cli/validate_config.py:
```diff
@@ -0,0 +1,56 @@
+@cover_packit_exception
+def validate_config(config, path_or_url):
+    """
+    Display PackageConfig validation.
```
Sorry I didn't notice this before:

```suggestion
    Validate Packit configuration.
```

This is a user facing text, while `PackageConfig` is an implementation detail. Users don't know about it, so it would be just confusing.

TomasTomecek commented:
> > > Is this the data to be validated? Shouldn't this be self.package_config instead?
> > 
> > 
> > I'm not sure if I understand. I renamed self.package_config to config because lines underneath were too long. I refactored validated dictionary into variable to solve this problem in latest commit.
> > Everything else should be fixed.
> 
> Thanks! Might be that I missed that renaming man_shrugging
> 
> I have concern regarding that mapping in `validate_primitive()` (which probably also applies to the other `validate_*` methods: if there is a change in `PackageConfig` this mapping will need to be also updated.
> 
> One way to overcome this could be to implement `get()` on `PackageConfig`, like:
> 
> ```python
>      def get(self, value, default=None):
>          return self.__dict__.get(value, default)
> ```
> 
> This way, the mapping wouldn't be needed (and we wouldn't be required to update it in the future when `PackageConfig` changes) and you could just call `validate(self.package_config)`.
> 
> @TomasTomecek how do you feel about the above?

Sounds good; even though I hate accessing self.__dict__, I'd say what you propose is a good solution

TomasJani commented:
> I have concern regarding that mapping in validate_primitive() (which probably also applies to the other validate_* methods: if there is a change in PackageConfig this mapping will need to be also updated.

I hope, that now it is working as you thought. The get function you mentioned was not enough, marsmallow checks type and request Mapping type for validation. So I extended config classes, that suppose to be validated (with Mapping abstract class) and added methods.

TomasTomecek commented on packit/api.py:
```diff
@@ -786,3 +787,54 @@ def clean(self):
         # command handlers have nothing to clean
         logger.debug("PackitAPI.cleanup (there are no objects to clean)")
+
+    def validation(self) -> str:
```
the common practice in python is to use verbs in method names instead of nouns, so this would be `validate`

TomasTomecek commented on packit/api.py:
```diff
@@ -786,3 +787,54 @@ def clean(self):
+        return index_output
+
+    @staticmethod
+    def validation_format_field_output(
```
this is an overkill to have a method for formatting a string, and it's only used in a single place - please just format string in-place above

TomasTomecek commented on packit/config/package_config.py:
```diff
@@ -253,6 +273,41 @@ def __eq__(self, other: object):
+
+    def __iter__(self):
+        package_dict = self.__dict__
+        del package_dict["_downstream_project_url"]
+        del package_dict["dist_git_clone_path"]
```
with this you are removing the attributes for good, since self.__dict__ is mutable; also why they need to be removed?

The best practice here is to copy self.__dict__

```python
package_dict = deepcopy(self.__dict__)

TomasTomecek commented on packit/config/package_config.py:
```diff
@@ -253,6 +273,41 @@ def __eq__(self, other: object):
+        setattr(self, key, value)
+
+    def __getitem__(self, key, default=None):
+        if key == "dist_git_url":
```
?

TomasTomecek commented on packit/cli/validate_config.py:
```diff
@@ -0,0 +1,56 @@
+from packit.cli.utils import get_packit_api
+from packit.config import pass_config, get_context_settings
+
+logger = logging.getLogger(__file__)
```
```suggestion
logger = logging.getLogger('packit')
```

`__file__` gives us the wrong longer and nothing gets printed

lachmanfrantisek replied:
Don't we want to use `__name__` instead?

```suggestion
logger = logging.getLogger(__name__)
```

TomasTomecek replied:
I actually fixed these in #857

TomasJani commented:
Sorry about this, I was just figuring out why it's printing unnecessary errors.  
>  def __getitem__(self, key, default=None):
>         if key == "dist_git_url":
> 
> del package_dict["_downstream_project_url"]
> del package_dict["dist_git_clone_path"]

Instead I added fields for _downstream_project_url and dist_git_clone_path into PackageConfigSchema so they would not be classified as unknown fields. 

Everything else should be okey.

TomasTomecek commented:
Sadly this doesn't work when I run it on packit's config:
```python
Traceback (most recent call last):
  File "/home/tt/g/user-cont/packit/packit/cli/utils.py", line 61, in covered_func
    func(config=config, *args, **kwargs)
  File "/home/tt/g/user-cont/packit/packit/cli/validate_config.py", line 55, in validate_config
    output = api.validate()
  File "/home/tt/g/user-cont/packit/packit/api.py", line 794, in validate
    schema_errors = PackageConfigValidation(self.package_config).validate()
  File "/home/tt/g/user-cont/packit/packit/config/package_config.py", line 213, in validate
    return PackageConfigSchema().validate(self.package_config)
  File "/usr/lib/python3.8/site-packages/marshmallow/schema.py", line 746, in validate
    self._do_load(data, many=many, partial=partial, postprocess=False)
  File "/usr/lib/python3.8/site-packages/marshmallow/schema.py", line 782, in _do_load
    processed_data = self._invoke_load_processors(
  File "/usr/lib/python3.8/site-packages/marshmallow/schema.py", line 1017, in _invoke_load_processors
    data = self._invoke_processors(
  File "/usr/lib/python3.8/site-packages/marshmallow/schema.py", line 1147, in _invoke_processors
    data = processor(data, many=many, **kwargs)
  File "/home/tt/g/user-cont/packit/packit/schema.py", line 339, in ordered_preprocess
    data = self.add_defaults_for_jobs(data, **kwargs)
  File "/home/tt/g/user-cont/packit/packit/schema.py", line 348, in add_defaults_for_jobs
    for job in data.get("jobs", []):
AttributeError: 'JobConfig' object has no attribute 'setdefault'
```

already looking into it

TomasTomecek commented:
Hm, so now that I play with this, I think that my work #859 changed config handling significantly which results in odd output:
```
$ packit validate-config
Input directory is an upstream repository.
.packit.yaml does not pass validation:
* field upstream_ref: Field may not be null.
* field actions: Unknown action(s) provided: [<ActionName.create_archive: 'create-archive'>, <ActionName.get_current_version: 'get-current-version'>]
* field allowed_gpg_keys: Field may not be null.
* field create_tarball_command: Field may not be null.
* field jobs has incorrect values:
** field 0 has incorrect values:
*** value at index specfile_path: Field may not be null.
*** value at index upstream_ref: Field may not be null.
*** value at index actions: Unknown action(s) provided: [<ActionName.create_archive: 'create-archive'>, <ActionName.get_current_version: 'get-current-version'>]
*** value at index downstream_package_name: Field may not be null.
*** value at index allowed_gpg_keys: Field may not be null.
*** value at index trigger: Enum name must be string
*** field metadata has incorrect values:
**** value at index _schema: Invalid input type.
*** value at index create_tarball_command: Field may not be null.
*** value at index job: Missing data for required field.
*** value at index type: Unknown field.
** field 1 has incorrect values:
*** value at index specfile_path: Field may not be null.
*** value at index upstream_ref: Field may not be null.
*** value at index actions: Unknown action(s) provided: [<ActionName.create_archive: 'create-archive'>, <ActionName.get_current_version: 'get-current-version'>]
*** value at index downstream_package_name: Field may not be null.
*** value at index allowed_gpg_keys: Field may not be null.
*** value at index trigger: Enum n
...
```

@jpopelka @sakalosj @csomh @lbarcziova @lachmanfrantisek would you have suggestions how to move forward here?

I'd personally say that #744 would help immensely.

@TomasJani this is what worked for me:
```diff
diff --git a/packit/schema.py b/packit/schema.py
index 2fbff3f..a96ba53 100644
--- a/packit/schema.py
+++ b/packit/schema.py
@@ -231,6 +231,10 @@ class JobMetadataSchema(MM23Schema):
 
     @pre_load
     def ordered_preprocess(self, data, **_):
+        if isinstance(data, JobMetadataConfig):
+            # this happens with the validate-config command when defaults are already applied
+            # this happens with the validate-config command when defaults are already applied
+            return data
         for key in ("dist-git-branch", "dist_git_branch"):
             if key in data:
                 logger.warning(
@@ -344,6 +348,9 @@ class PackageConfigSchema(CommonConfigSchema):
         """
         add all the fields (except for jobs) to every job so we can process only jobconfig in p-s
         """
+        if isinstance(data, (PackageConfig, JobMetadataConfig)):
+            # this happens with the validate-config command when defaults are already applied
+            return data
         for job in data.get("jobs", []):
             for k, v in data.items():
                 if k == "jobs":
```

csomh commented:
> @jpopelka @sakalosj @csomh @lbarcziova @lachmanfrantisek would you have suggestions how to move forward here?
> 
> I'd personally say that #744 would help immensely.

TBH, I think @sakalosj has a point in the discussion at #744: there is a reason schema objects are kept separate from the objects they translate to/from in Marshmallow, and that is: they do different things. I'm pretty sure if merging them would result in better readability or easier maintenance, the Marshmallow docs would mention it.

The discussion on that issue probably also can help us understand what's wrong here: it seems that so far we completely missed (apologies for that!) that `PackageConfigValidation` works on the config objects themselves, instead of working on the configuration data (a.k.a the YAML or better said the dictionary parsed from it).

This is why inheriting from Mappings was required. But I think this is actually not the right way to use Marshmallow or to do the schema validation: all the `validation` command at this point should do is `yaml.load` the configuration file and `PackageConfigSchema().validate()` it. That's it.

There is no need to deserialize the configuration into the configuration objects, as those objects wouldn't be used anyways at this point. (Hope my understanding and memory about this PR handling only the schema validation part—not the dynamic value checks—is still valid.)

This would check that the schema is valid, but would not actually create the config objects, so things happening at deserialization wouldn't interfere with schema validation.

I hope the above is clear, if not please shout! :smiley:

TomasJani commented:
Thank you for suggestions.
I adjusted validation accordingly.
Let me know if something is not properly done.

TomasTomecek commented:
/packit build
